File: add_two_numbers/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbers(l1, l2, 0);
    }
    
    private ListNode addTwoNumbers(ListNode l1, ListNode l2, int carryOver) {
        if (l1 == null && l2 == null && carryOver == 0) {
            return null;
        }
        if (l1 == null && l2 == null) {
            return new ListNode(carryOver);
        }
        if (l1 == null) {
            carryOver += l2.val;
            l2.val = carryOver % 10;
            carryOver /= 10;
            l2.next = addTwoNumbers(l1, l2.next, carryOver);
            return l2;
        }
        if (l2 == null) {
            carryOver += l1.val;
            l1.val = carryOver % 10;
            carryOver /= 10;
            l1.next = addTwoNumbers(l1.next, l2, carryOver);
            return l1;
        }
        
        carryOver += l1.val + l2.val;
        l1.val = carryOver % 10;
        carryOver /= 10;
        l1.next = addTwoNumbers(l1.next, l2.next, carryOver);
        return l1;
    }
}

File: add_two_numbers/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Add_Two_Numbers {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1 == NULL && l2 == NULL) {
            return NULL;
        }
        if (l1 == NULL) {
            return l2;
        }
        if (l2 == NULL) {
            return l1;
        }
        
        ListNode* dummy = new ListNode(0);
        dfs(l1, l2, 0, dummy);
        return dummy->next;
    }

private:
    void dfs(ListNode* l1, ListNode* l2, int carry_over, ListNode* curr_node) {
        if (l1 == NULL && l2 == NULL && carry_over == 0) {
            return;
        }
        if (l1 == NULL && l2 == NULL) {
            curr_node->next = new ListNode(carry_over);
            return;
        }
        if (l1 == NULL) {
            carry_over += l2->val;
            curr_node->next = new ListNode(carry_over % 10);
            dfs(l1, l2->next, carry_over / 10, curr_node->next);
            return;
        }
        if (l2 == NULL) {
            carry_over += l1->val;
            curr_node->next = new ListNode(carry_over % 10);
            dfs(l1->next, l2, carry_over / 10, curr_node->next);
            return;
        }
        
        carry_over += l1->val + l2->val;
        curr_node->next = new ListNode(carry_over % 10);
        dfs(l1->next, l2->next, carry_over / 10, curr_node->next);        
    }
};